src/others/xml/parsers/lv_xml_obj_parser.c

```diff
@@ -0,0 +1,81 @@
+/**
+ * @file lv_xml_obj_parser.c
+ *
+ */
+
+/*********************
+ *      INCLUDES
+ *********************/
+#include "lv_xml_obj_parser.h"
+#if LV_USE_XML
+
+#include "../../../lvgl.h"
+#include "../../../lvgl_private.h"
+
+/*********************
+ *      DEFINES
+ *********************/
+
+/**********************
+ *      TYPEDEFS
+ **********************/
+
+/**********************
+ *  STATIC PROTOTYPES
+ **********************/
+static int32_t size_process(const char * value_str);
+
+/**********************
+ *  STATIC VARIABLES
+ **********************/
+
+/**********************
+ *      MACROS
+ **********************/
+
+/**********************
+ *   GLOBAL FUNCTIONS
+ **********************/
+
+void * lv_xml_obj_process(lv_xml_parser_state_t * state, const char ** attrs)
+{
+    void * item = lv_obj_create(lv_xml_state_get_parent(state));
+
+    lv_obj_xml_apply_attrs(state, item, attrs);
+
+    return item;
+}
+
+void lv_obj_xml_apply_attrs(lv_xml_parser_state_t * state, lv_obj_t * obj, const char ** attrs)
+{
+    for(int i = 0; attrs[i]; i += 2) {
+        const char * name = attrs[i];
+        const char * value = attrs[i + 1];
+
+        if(lv_streq("x", name)) lv_obj_set_x(obj, size_process(value));
+        if(lv_streq("y", name)) lv_obj_set_y(obj, size_process(value));
+        if(lv_streq("width", name)) lv_obj_set_width(obj, size_process(value));
+        if(lv_streq("height", name)) lv_obj_set_height(obj, size_process(value));
+        if(lv_streq("align", name)) lv_obj_set_align(obj, lv_xml_align_string_to_enum_value(value));
+        if(lv_streq("style_radius", name)) lv_obj_set_style_radius(obj, lv_xml_atoi(value), 0);
+        if(lv_streq("style_bg_color", name)) lv_obj_set_style_bg_color(obj, lv_color_hex(lv_xml_strtol(value, NULL, 16)), 0);
+        if(lv_streq("style_text_color", name)) lv_obj_set_style_text_color(obj, lv_color_hex(lv_xml_strtol(value, NULL, 16)),
+                                                                               0);
+        if(lv_streq("styles", name)) lv_xml_style_add_to_obj(state, obj, value);
```

Comment:
The code is repeated in property, subject and here. I don't think it's a good idea to write a separate lv_xml_xxx_parser for every widget.

Reply:
First of all, thank you for looking into it. 

By property do you mean lv_obj_property? If so I see what you mean and it'd be great to have a single place for these kind of codes. However I found obj_property limited in some cases (see [here](https://github.com/lvgl/lvgl/issues/6915#issuecomment-2383074845))

So maybe we should approach it the other way around. Build obj_property on top of the parsers somehow.

Reply:
>  it'd be great to have a single place for these kind of codes

Yes. We can make it really `unified` API.

I'm aware of the limitation of single value per property.
I'm thinking of the solution to support recursive(or combined) value like property of properties.

```c

typedef struct {
    lv_prop_id_t id; /* Reserve 3 bits to indicate it's a combined property, and how many property values are inside. */
    union {
     lv_property_t * sub_properties[]; /* Pointers to an array of lv_property_t */
    ....
    };
} lv_property_t;
```

I'm unfortunately don't have the time to try it out. It would be really great if it can be used in both subject, xml and language bindings.

By doing so, user implemented widgets only have to provide the property table, all subject, xml, and bindings are able to set/get property use the general property API.

Reply:
It seems like a great idea,  but I've just realized something. By using the property API we could replace this line
```c
if(lv_streq("x", name)) lv_obj_set_x(obj, size_process(value));
```
with
```c
if(lv_streq("x", name))  lv_obj_set_property(obj, (lv_property_t){.id=LV_PROPERTY_OBJ_X, .num=size_process(value)});
```

It can be structured differently but in the end `"x"` needs to mapped to `LV_PROPERTY_OBJ_X` and the string value needs to be processed as well.

Once we need to map `"x"` to something we can map it directly to `lv_obj_set_x`. In this sense using the property API seems like an extra abstraction layer. That is to convert the attribute and value to a property descriptor, call the property API which calls the LVGL API, instead of calling the LVGL API directly. Now I can not see the advantage of the property API here. Let me know if I miss something. 

> I'm unfortunately don't have the time to try it out. 

Could you or @FASTSHIFT just try out the example, to make sure that it works on others' computer too. :slightly_smiling_face: You just need to enable an FS driver with letter `'A'`.

Reply:
`lv_prop_id_t lv_obj_property_get_id(const lv_obj_t * obj, const char * name);` is used to get the property ID, thus no need to string compare it manually.

The name to ID map is automatically generated by script `lvgl/scripts/properties.py`.
See example of `lvgl/src/widgets/property/lv_label_properties.c`
**User only need to write once definition of property ID in widget implementation(E.g `lv_image.c`).** Then all the remaining is automatic. No need to write porting layer for xml manually.

For xml and other bindings, an API to get the value from description will be needed. E.g

`lv_xml_convert_property_value(lv_prop_id_t id, const char* value_string)`

I do in this way in lua:
```c
LUALIB_API lv_property_t luavgl_toproperty(lua_State *L, int idx,
                                           lv_prop_id_t id)
{
  lv_property_t prop = {0};
  prop.id = id;
  prop.num = 0;

  int type = LV_PROPERTY_ID_TYPE(id);
  switch (type) {
  case LV_PROPERTY_TYPE_INT:
    prop.num = luavgl_tointeger(L, idx);
    return prop;
  case LV_PROPERTY_TYPE_COLOR:
    prop.color = luavgl_tocolor(L, idx);
    return prop;
...
  default:
    break;
  }

  prop.id = LV_PROPERTY_ID_INVALID;
  return prop;
}
```

I imagine xml will need this:
```xml
<my_h3 text="$btn_text" align="center" color="#orange"/>
```
key: color
value: "#orange"

```c
id = lv_obj_property_get_id(btn, "color")
lv_property_t value = lv_xml_convert_property_value(id, "#orange")
lv_obj_set_property(btn, &value);
```

Reply:
Oh, I see. Sounds promising.

Do you have any idea for converting enums, e.g. `"center"` to `LV_ALIGN_CENTER`.

Reply:
`align` parameter has int type of property value. Using `"center"` string for convention will need to convert to int by xml processor itself. Similarly for color property with convention of `#orange` and `#F12` etc.

E.g.
```c
LUALIB_API lv_color_t luavgl_tocolor(lua_State *L, int idx)
{
  lv_color_t color = {0};
  if (lua_type(L, idx) == LUA_TSTRING) {
    /* support #RGB and #RRGGBB */
    const char *s = lua_tostring(L, idx);
    if (s == NULL) {
      luaL_error(L, "unknown color");
      return color;
    }

    int len = lv_strlen(s);
    if (len == 4 && s[0] == '#') {
      /* #RGB */
      int r = to_int(s[1]);
      r |= r << 4;
      int g = to_int(s[2]);
      g |= g << 4;
      int b = to_int(s[3]);
      b |= b << 4;
      color = lv_color_make(r, g, b);
    } else if (len == 7 && s[0] == '#') {
      /* #RRGGBB */
      int r = (to_int(s[1]) << 4) | to_int(s[2]);
      int g = (to_int(s[3]) << 4) | to_int(s[4]);
      int b = (to_int(s[5]) << 4) | to_int(s[6]);
      color = lv_color_make(r, g, b);
    } else {
      luaL_error(L, "unknown color format");
      return color;
    }
  } else {
    color = lv_color_hex(luavgl_tointeger(L, idx)); /* make to lv_color_t */
  }

  return color;
}
```

Reply:
I see. Probably we can generate it somehow too, as it's a quite straight forward mapping. Anyway, at this point I'd like to see the editor working as a whole first, and after that we can get beck to this. We can change this XML parser part later too without modifying any public APIs.